lightning/src/util/events.rs

```diff
@@ -147,6 +148,14 @@ pub enum Event {
 #[cfg(test)]
 		error_data: Option<Vec<u8>>,
 	},
+	/// Indicates an outbound MPP payment partially failed. Probably some intermediary node dropped
+	/// something. You may wish to retry this portion of the payment with a different path.
+	MPPFragmentFailed {
+		/// The hash which was given to ChannelManager::send_payment.
+		payment_hash: PaymentHash,
+		/// The path that failed.
+		payment_path: Vec<RouteHop>,
```

Comment:
I'm working on a PR that adds the path to `PaymentFailed` for scoring purpose. It will also include the index of the failed hop. Given that, do you think we need a separate event? Could we add a marker for multi-path payments if needed?

Reply:
I don't have a strong opinion. Matt pointed out in [this comment](https://github.com/rust-bitcoin/rust-lightning/pull/1053#discussion_r694306740) that there could be value in users knowing whether to retry a portion of a payment vs. retrying the entire payment. @TheBlueMatt I tend to agree that this might make `MPPFragmentFailed` a bit redundant, thoughts?

Reply:
I could break out the relevant commits into their own PR if it helps. I think from a retry perspective, though, we'd need to know the part amount unless `InvoicePayer` tracks that.

Reply:
Discussed this one offline, but we just need to make sure that we expose to users somehow the distinction between a *payment* failing and an MPP fragment failing, as they have significant implication for what the user does in response. We *really* don't want users to think that a `PaymentFailed` event implies the payment failed, when it was only one mpp fragment that failed.

Reply:
Drafted https://github.com/rust-bitcoin/rust-lightning/pull/1077